src/language/patterns.md

```diff
@@ -0,0 +1,358 @@
+
+for (var MapEntry(key: key, value: count) in hist.entries) {
+  print('$key occurred $count times');
+}
+```
+
+The object pattern checks that `hist.entries` has the named type `MapEntry`,
+and then recurses into the named field subpatterns `key` and `value`.
+It calls the `key` getter and `value` getter on the `MapEntry` in each iteration,
+and binds the results to local variables `key` and `count`, respectively.
+
+Binding the result of a getter call to a variable of the same name is a common
+use case, so object patterns can also infer the getter name from the
+[variable subpattern][variable]. This allows you to simplify the variable pattern
+from something redundant like `key: key` to just `:key`:
+
+```dart
+for (var MapEntry(:key, value: count) in hist.entries) { 
+  print('$key occurred $count times');
+}
+```
+
+## Uses cases for patterns
+
+The [section above](#places-patterns-can-appear)
+describes _how_ you can use patterns in Dart code constructs, 
+sometimes with interesting uses cases as examples (like [swapping](#variable-assignment)
+the values of two variables, or
+[destructuring key-value pairs](#for-and-for-in-loops)
+in a map). 
+
+This section describes even more use cases, answering _when and why_ you might
+want to use patterns, the kinds of problems they solve, and the idioms they're 
+best suited for.
+
+### Destructuring multiple returns
+
+[Records][] allow aggregating and returning multiple values from a single function
+call. Patterns add the ability to destructure a record's fields
+directly into local variables, inline with the function call.
+
+Instead of individually declaring new local variables for each record field,
+like this:
+
+```dart
+main() {
+  var info = userInfo(json);
+  var name = info.$1;
+  var age = info.$2;
+}
+```
+
+You can destructure the fields of a record that a function returns into local
+variables using a [variable declaration](#variable-declaration) or
+[assigment pattern](#variable-assignment), and a record pattern as its subpattern:
+
+```dart
+main() {
+  var (name, age) = userInfo(json);
+}
+```
+
+### Destructuring class instances
+
+[Object patterns][object] match against named object types, allowing
+you to destructure their data using the getters the object's class already exposes.
+
+To destructure an instance of a class, use the named type, followed by the properties
+to destructure enclosed in parenthesis:
+
+```dart
+main() {
+  final Foo myFoo = Foo(one: 'one', two: 2);
+  var Foo(:one, :two) = myFoo;
+  print('one $one, two $two');
+}
+```
+
+### Algebraic datatypes 
+
+Object destructuring and switch cases are conducive to writing
+code in an [algebraic datatype][] style: 
+when you have a family of related types and an operation that needs
+specific behavior for each type.
+
+Instead of implementing the operation as an instance method for every respective type,
+you can keep the operations in a single function, switching over subtypes for
+variations in behavior:
```

Comment:
**issue:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading). One 29 word sentence to three ~10 word sentences.
```suggestion
This provides an alternative to each type having its own instance method.
Keep the operations, or instance methods, in one function.
Use a switch statement to to through subtypes for behavior variations.
```

Reply:
I liked the way this sentence flowed, so I cut some words out to get it down to 26 per GSG instead of restructuring it 😬 

> Instead of implementing the operation as an instance method for every type,
keep the operations in a single function, switching over subtypes for
behavior variations: